### ISSUE TITLE
A false positive `-Wmissing-field-initializers` for anonymous unions

### ISSUE THREAD

**Issue by Fznamznon** [llvm/llvm-project#70444]
https://reviews.llvm.org/D157879 introduced a false positive for anonymous union members:

```
struct A {
    int m;
    union { int n = 0; };
};

A a = A{.m = 0};
```

now produces a false positive warning saying that the anonymous union member in A is uninitialized.

Thanks @zygoloid for reporting this.

**Comment by Fznamznon** [comment]
I gave it some thought.... Actually, should clang warn about it at all?
https://reviews.llvm.org/D157879 only enabled warnings for designators, the problem has been there before, can be reproduced without designators - https://godbolt.org/z/7M5o9vWY4. Note that gcc reports this as well.
If we read "missing initializer" as not explicitly provided in the list then the warning is correct. 
If we choose to not report this particular case, should we also not report something like:

```
union B {
    int a = 1;
    float b;
};

struct A {
    int m;
    B BB;
};
void foo() {
  A a = {.m = 1};
}

```
https://godbolt.org/z/7ev8Y95Y7

or should we not report something like this:

```
struct A {
    int m;
    union { int a; union { short b; float cc = 1; }; };

};
void foo() {
  A a = {.m = 1};
}
```
The latter asserts clang atm though.

cc @AaronBallman @shafik for opinion.

**Comment by shafik** [comment]
It seems like there is a lot of debate on this diagnostic: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96868#c3

I can see points on both sides of this discussion. There seems to be sufficient confusion that maybe we should have a separate diagnostic.

**Comment by zygoloid** [comment]
> If we read "missing initializer" as not explicitly provided in the list then the warning is correct.

Sure, but in general the warning does not warn on members with default member initializers, because in that case the initializer isn't missing -- it's present, in the class definition. In this case, there is a default member initializer.

> If we choose to not report this particular case, should we also not report something like:
> 
> ```
> union B {
>     int a = 1;
>     float b;
> };
> 
> struct A {
>     int m;
>     B BB;
> };
> void foo() {
>   A a = {.m = 1};
> }
> ```

This situation is different, because the author of struct `A` did not explicitly say that the initializer of `BB` is optional. In the original example, they did. They could write `B BB = {};` if that's what they mean. In the original example, they can't.

> or should we not report something like this:
> 
> ```
> struct A {
>     int m;
>     union { int a; union { short b; float cc = 1; }; };
> 
> };
> void foo() {
>   A a = {.m = 1};
> }
> ```

We shouldn't diagnose that under this warning flag. But we should warn that this is not valid C++ (anonymous unions can't be nested) and perhaps warn that it's pointless to nest them (flattening the two unions into one would change nothing).

**Comment by Fznamznon** [comment]
Duplicate.
